### PR TITLE
dyn scoped methods

### LINKED ISSUE #9: Consider adding methods under field of contract object
Currently the API attaches functions for each function in the ABI based on name, but this can cause issues if the function name overlaps with one of the functions we currently do or will support in the future.

Might be the safest to just have a `methods` field on the contract. Worth getting input from someone with more JS exp than me.

Benefits:
- Improves typing for the `AnyContract` object. Currently there are no types available since the wildcard on the fields of the contract overwrites the TS types
- Avoids overlap and potentially buggy behaviour

Cons:
- Can't call the method on the contract object directly (slightly less ergonomic):
```ts
contract.add([1, 2], [3, 4], 5);

// Would be
contract.methods.add([1, 2], [3, 4], 5);
```

### PR DESCRIPTION
**Pull request by miraclx**
Resolves #9.

Defines the following interface:

```js
> contract = new Contract(connection, "dev-1661966288541-80307357536154", abiSchema);
Contract {
  methods: ContractMethods {
    add: [Function: ContractMethod[add]],
    add_borsh: [Function: ContractMethod[add_borsh]],
    add_callback: [Function: ContractMethod[add_callback]],
    no_params: [Function: ContractMethod[no_params]]
  }
}
```

```js
> contract.methods
ContractMethods {
  add: [Function: ContractMethod[add]],
  add_borsh: [Function: ContractMethod[add_borsh]],
  add_callback: [Function: ContractMethod[add_callback]],
  no_params: [Function: ContractMethod[no_params]]
}
```

```js
> contract.methods.add
[Function: ContractMethod[add]]
```

```js
> contract.methods.add([1, 2], [3, 4])
ContractMethodInvocation {
  view: [AsyncFunction: ContractMethod[add].view]
}
```

```js
> await contract.methods.add([1, 2], [3, 4]).view()
[4, 6]
```

```js
> contract.methods.add([1, 2], [3, 4]).arguments
[[1, 2], [3, 4]]
```

```js
> contract.methods.add([1, 2], [3, 4]).contract
Contract { [... cyclic ...] }
```

```js
> contract.methods.add([1, 2], [3, 4]).method
{ "name": "add", doc: ... }
```

### CODE REVIEW

File: src/contract.ts [hunk 1]
```diff
@@ -166,74 +149,131 @@ function serializeArgs(
...
+    /**
+     * @param contract NEAR Contract object
+     */
+    constructor(contract: Contract) {
+        if (!contract.abi)
+            throw new Error("Can't create ContractMethods without ABI");
+        // Create method on this contract object to be able to call methods.
+        for (const fn of contract.abi.body.functions) {
+            const handler = (...args): ContractMethod => {
+                return new ContractMethod(contract, fn, args);
```
Comment (itegulov) [inline src/contract.ts]: Any reason to create a new instance on every invocation?

File: src/contract.ts [hunk 2]
```diff
@@ -166,74 +149,131 @@ function serializeArgs(
...
+                // not be ideal API.
+                return callInternal(
+                    account,
+                    contract.contractId,
+                    fn.name,
+                    serializeArgs(fn.name, args, fn.params),
+                    opts
+                );
+            };
+            Object.defineProperty(this.view, 'name', {
```
Comment (itegulov) [inline src/contract.ts]:
```suggestion
            Object.defineProperty(this.call, 'name', {
```

File: src/contract.ts (hunk 1 above)
Comment (miraclx) [reply src/contract.ts]: Could probably be more succinctly named, but this is effectively an object that that associates arguments to a method.

```js
> let add_method: ContractMethod = contract.methods.add([1, 2], [3, 4])

> add_method.arguments
[[1, 2], [3, 4]]

> await add_method.view()
[4, 6]
```

Got any ideas for an alternative name?

Comment (itegulov) [reply src/contract.ts]: Ah, I see, missed the `args` part.

Maybe `ContractMethodInvocation`? A bit wordy though

**Review by itegulov — APPROVED** [review]
Regardless what we choose for the name LGTM!

File: src/contract.ts (hunk 1 above)
Comment (miraclx) [reply src/contract.ts]: Yeah, that works, if needed, we can revise the name later. Thanks!
